flake8_trio/visitors/visitor91x.py

```diff
@@ -204,123 +340,154 @@ def visit_Try(self, node: ast.Try):
             # very bad)
             if not any(
                 h.type is None
-                or (isinstance(h.type, ast.Name) and h.type.id == "BaseException")
+                or (isinstance(h.type, cst.Name) and h.type.value == "BaseException")
                 for h in node.handlers
             ):
-                self.uncheckpointed_statements.update(added)
+                self.uncheckpointed_statements.update(self.try_state.added)
+
+    def leave_Try_finalbody(self, node: cst.Try):
+        if node.finalbody:
+            self.uncheckpointed_statements.difference_update(self.try_state.added)
 
-            self.visit_nodes(node.finalbody)
-            self.uncheckpointed_statements.difference_update(added)
+    def leave_Try(self, original_node: cst.Try, updated_node: cst.Try) -> cst.Try:
+        self.restore_state(original_node)
+        return updated_node
 
-    # valid checkpoint if both body and orelse checkpoint
-    def visit_If(self, node: ast.If | ast.IfExp):
+    def leave_If_test(self, node: cst.If | cst.IfExp) -> None:
         if not self.async_function:
             return
-        # visit condition
-        self.visit_nodes(node.test)
-        outer = self.uncheckpointed_statements.copy()
+        self.save_state(node, "uncheckpointed_statements", copy=True)
 
-        # visit body
-        self.visit_nodes(node.body)
-        body_outer = self.uncheckpointed_statements
-
-        # reset to after condition and visit orelse
-        self.uncheckpointed_statements = outer
-        self.visit_nodes(node.orelse)
+    def leave_If_body(self, node: cst.If | cst.IfExp) -> None:
+        if not self.async_function:
+            return
 
-        # union of both branches is the new set of unhandled entries
-        self.uncheckpointed_statements.update(body_outer)
+        # restore state to after test, saving current state instead
+        (
+            self.uncheckpointed_statements,
+            self.outer[node]["uncheckpointed_statements"],
+        ) = (
+            self.outer[node]["uncheckpointed_statements"],
+            self.uncheckpointed_statements,
+        )
 
-    # inline if
-    visit_IfExp = visit_If
+    def leave_If(self, original_node: cst.If, updated_node: cst.If) -> cst.If:
+        if self.async_function:
+            # merge current state with post-body state
+            self.uncheckpointed_statements.update(
+                self.outer[original_node]["uncheckpointed_statements"]
+            )
+        return updated_node
+
+    # libcst calls attributes in the order they appear in the code, so we manually
+    # rejig the order here
+    def visit_IfExp(self, node: cst.IfExp) -> bool:
```

Comment:
could in ast be handled the exact same way, but not requires some slight rejigging